File: src/main/java/core/procedures/vectors/VectorToList.java
package core.procedures.vectors;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMCons;
import core.scm.SCMMutableVector;
import core.scm.SCMVector;

public final class VectorToList extends AFn {

  public VectorToList() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{SCMVector.class}).build());
  }

  @Override
  public String getName() {
    return "vector->list";
  }

  @Override
  public SCMCons apply1(Object arg) {
    return vectorToList((SCMMutableVector)arg);
  }

  public static SCMCons vectorToList(SCMMutableVector v) {
    return SCMCons.list((v).getArray());
  }
}

File: src/main/java/core/procedures/vectors/VectorToList.kt
package core.procedures.vectors

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Vector

class VectorToList : AFn<Vector, List<*>>(name = "vector->list", isPure = true, arity = Exactly(1),
                                          mandatoryArgsTypes = arrayOf(Vector::class.java)) {

    override operator fun invoke(arg: Vector) = arg.toList()
}
